lib/internal/test_runner/yaml_parser.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+const {
+  codes: {
+    ERR_TEST_FAILURE,
+  }
+} = require('internal/errors');
+const AssertionError = require('internal/assert/assertion_error');
+const {
+  ArrayPrototypeJoin,
+  Error,
+  Number,
+  NumberIsNaN,
+  RegExpPrototypeExec,
+  StringPrototypeEndsWith,
+  StringPrototypeRepeat,
+  StringPrototypeSlice,
+  StringPrototypeStartsWith,
+  StringPrototypeSubstring,
+} = primordials;
+
+const kYamlKeyRegex = /^(\s+)?(\w+):(\s)+([>|][-+])?(.*)$/;
+const kStackDelimiter = '    at ';
+
+function reConstructError(parsedYaml) {
+  if (!('error' in parsedYaml)) {
+    return parsedYaml;
+  }
+  const isAssertionError = parsedYaml.code === 'ERR_ASSERTION' ||
+    'actual' in parsedYaml || 'expected' in parsedYaml || 'operator' in parsedYaml;
+  const isTestFauilure = parsedYaml.code === 'ERR_TEST_FAILURE' || 'failureType' in parsedYaml;
+  const stack = parsedYaml.stack ? kStackDelimiter + ArrayPrototypeJoin(parsedYaml.stack, `\n${kStackDelimiter}`) : '';
+
+  // eslint-disable-next-line no-restricted-syntax
+  let error = new Error(parsedYaml.error);
+  error.stack = stack;
+  error.code = parsedYaml.code;
+
+  if (isAssertionError) {
+    error = new AssertionError({
+      message: parsedYaml.error,
+      actual: parsedYaml.actual,
+      expected: parsedYaml.expected,
+      operator: parsedYaml.operator
+    });
+    error.stack = stack;
+  }
+
+  if (isTestFauilure) {
+    error = new ERR_TEST_FAILURE(error, parsedYaml.failureType);
+    error.stack = stack;
+  }
```

Comment:
```suggestion
  let error;
  
  if (isTestFauilure) {
    error = new ERR_TEST_FAILURE(error, parsedYaml.failureType);
  } else if (isAssertionError) {
    error = new AssertionError({
      message: parsedYaml.error,
      actual: parsedYaml.actual,
      expected: parsedYaml.expected,
      operator: parsedYaml.operator
    });
  } else {
    // eslint-disable-next-line no-restricted-syntax
    error = new Error(parsedYaml.error);
    error.code = parsedYaml.code;
  }

  error.stack = stack;
```

Reply:
@aduh95 @cjihrig this does not behave the same since there is  a possibility where `isTestFauilure === isAssertionError === true` and we want to receive a `new ERR_TEST_FAILURE(new AssertionError())`
see my additional commit